nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/common/AbstractMQTTProcessor.java

```diff
@@ -0,0 +1,342 @@
+import org.apache.nifi.processor.util.StandardValidators;
+import org.apache.nifi.ssl.SSLContextService;
+import org.eclipse.paho.client.mqttv3.IMqttClient;
+import org.eclipse.paho.client.mqttv3.MqttClient;
+import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
+import org.eclipse.paho.client.mqttv3.MqttException;
+import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Properties;
+
+import static org.apache.nifi.processors.mqtt.common.MqttConstants.ALLOWABLE_VALUE_CLEAN_SESSION_FALSE;
+import static org.apache.nifi.processors.mqtt.common.MqttConstants.ALLOWABLE_VALUE_CLEAN_SESSION_TRUE;
+import static org.apache.nifi.processors.mqtt.common.MqttConstants.ALLOWABLE_VALUE_MQTT_VERSION_310;
+import static org.apache.nifi.processors.mqtt.common.MqttConstants.ALLOWABLE_VALUE_MQTT_VERSION_311;
+import static org.apache.nifi.processors.mqtt.common.MqttConstants.ALLOWABLE_VALUE_MQTT_VERSION_AUTO;
+import static org.apache.nifi.processors.mqtt.common.MqttConstants.ALLOWABLE_VALUE_QOS_0;
+import static org.apache.nifi.processors.mqtt.common.MqttConstants.ALLOWABLE_VALUE_QOS_1;
+import static org.apache.nifi.processors.mqtt.common.MqttConstants.ALLOWABLE_VALUE_QOS_2;
+
+public abstract class AbstractMQTTProcessor extends AbstractSessionFactoryProcessor {
+
+    protected ProcessorLog logger;
+    protected IMqttClient mqttClient;
+    protected final Object mqttClientConnectLock = new Object();
+    protected volatile String broker;
+    protected volatile String clientID;
+    protected MqttConnectOptions connOpts;
+    protected MemoryPersistence persistence = new MemoryPersistence();
+    protected long maxTimeout;
+
+    public ProcessSessionFactory processSessionFactory;
+
+    public static final Validator QoSValidator = new Validator() {
+
+        @Override
+        public ValidationResult validate(String subject, String input, ValidationContext context) {
+            Integer inputInt = Integer.parseInt(input);
+            if (inputInt < 0 || inputInt > 2) {
+                return new ValidationResult.Builder().subject(subject).valid(false).explanation("QoS must be an integer between 0 and 2").build();
+            }
+            return new ValidationResult.Builder().subject(subject).valid(true).build();
+        }
+    };
+
+    public static final Validator BrokerValidator = new Validator() {
+
+        @Override
+        public ValidationResult validate(String subject, String input, ValidationContext context) {
+            try{
+                URI brokerURI = new URI(input);
+                if (!"".equals(brokerURI.getPath())) {
+                    return new ValidationResult.Builder().subject(subject).valid(false).explanation("the broker URI cannot have a path. It currently is:" + brokerURI.getPath()).build();
```

Comment:
Something like that (if possible). Just to avoid duplication of code. That's the only argument, unless using it will prove more complicated. Your call.